tracer/test/Datadog.Trace.IntegrationTests/Tagging/TraceContextPropertyTests.cs

```diff
@@ -0,0 +1,163 @@
+    [InlineData(" ", null)]                                    // whitespace
+    [InlineData("test", "test")]                               // normal
+    [InlineData("Caps and Spaces! 🐶", "Caps and Spaces! 🐶")] // adding this now so it fails later when we start normalizing this
+    public async Task Env(string before, string after)
+    {
+        using (var scope = CreateTrace(_tracer))
+        {
+            scope.Span.Context.TraceContext.Environment = before;
+        }
+
+        await _tracer.FlushAsync();
+        var traceChunks = _testApi.Wait();
+        var spans = traceChunks.SelectMany(s => s).ToArray();
+
+        spans.Should().HaveCount(4);
+
+        if (after is null)
+        {
+            spans.Should().OnlyContain(s => !s.Tags.ContainsKey("env"));
+        }
+        else
+        {
+            spans.Should().OnlyContain(s => s.Tags["env"] == after);
+        }
+    }
+
+    [Theory]
+    [InlineData(null, null)]                                   // null
+    [InlineData("", null)]                                     // empty
+    [InlineData(" ", null)]                                    // whitespace
+    [InlineData("test", "test")]                               // normal
+    [InlineData("Caps and Spaces! 🐶", "Caps and Spaces! 🐶")] // adding this now so it fails later when we start normalizing this
+    public async Task Version(string before, string after)
+    {
+        using (var scope = CreateTrace(_tracer))
+        {
+            scope.Span.Context.TraceContext.ServiceVersion = before;
+        }
+
+        await _tracer.FlushAsync();
+        var traceChunks = _testApi.Wait();
+        var spans = traceChunks.SelectMany(s => s).ToArray();
+
+        spans.Should().HaveCount(4);
+
+        if (after is null)
+        {
+            spans.Should().OnlyContain(s => !s.Tags.ContainsKey("version"));
+        }
+        else
+        {
+            spans.Should().OnlyContain(s => s.Tags["version"] == after);
+        }
+    }
+
+    [Theory]
+    [InlineData(null, null)]                                   // null
+    [InlineData("", null)]                                     // empty
+    [InlineData(" ", null)]                                    // whitespace
+    [InlineData("test", "test")]                               // normal
+    [InlineData("Caps and Spaces! 🐶", "Caps and Spaces! 🐶")] // adding this now so it fails later when we start normalizing this
+    public async Task Origin(string before, string after)
+    {
+        using (var scope = CreateTrace(_tracer))
+        {
+            scope.Span.Context.TraceContext.Origin = before;
+        }
+
+        await _tracer.FlushAsync();
+        var traceChunks = _testApi.Wait();
+        var spans = traceChunks.SelectMany(s => s).ToArray();
+
+        spans.Should().HaveCount(4);
+
+        if (after is null)
+        {
+            spans.Should().OnlyContain(s => !s.Tags.ContainsKey("version"));
```

Comment:
Should this be `origin`? Also, I think you can refactor these 3 tests to assert the same things in another function, just changing the tag key, if you think that would be worth it.

```suggestion
            spans.Should().OnlyContain(s => !s.Tags.ContainsKey("origin"));
```

Reply:
Oops, good catch, thanks! This is one of the tests I mentioned is hanging so I haven't been able to run it yet.

Reply:
Done.